File: implementations/python/pocketplus/mask.py
"""
POCKET+ mask update logic.

Implements CCSDS 124.0-B-1 Section 4 (Mask Update):
- Build Vector Update (Equation 6)
- Mask Vector Update (Equation 7)
- Change Vector Computation (Equation 8)

MicroPython compatible - no typing module imports.
"""

# Import for type hints only
if False:  # noqa: SIM108
    from pocketplus.bitvector import BitVector


def update_build(
    build: "BitVector",
    input_vec: "BitVector",
    prev_input: "BitVector",
    new_mask_flag: bool,
    t: int,
) -> None:
    """
    Update the build vector (CCSDS Equation 6).

    Build vector accumulates bits that have changed over time.
    When new_mask_flag is set, the build is used to replace the mask
    and is then reset to zero.

    Equation:
    - Bt = (It XOR It-1) OR Bt-1  (if t > 0 and new_mask_flag = 0)
    - Bt = 0                      (if t = 0 or new_mask_flag = 1)

    Args:
        build: Build vector to update (modified in place)
        input_vec: Current input vector It
        prev_input: Previous input vector It-1
        new_mask_flag: Whether a new mask is being established
        t: Current time step
    """
    if t == 0 or new_mask_flag:
        # Reset build to zero
        build.zero()
    else:
        # Bt = (It XOR It-1) OR Bt-1
        # Create temporary for changes
        changes = input_vec.copy()
        changes.xor(changes, prev_input)

        # Update build: build = changes OR build
        build.or_(build, changes)


def update_mask(
    mask: "BitVector",
    input_vec: "BitVector",
    prev_input: "BitVector",
    build_prev: "BitVector",
    new_mask_flag: bool,
) -> None:
    """
    Update the mask vector (CCSDS Equation 7).

    The mask tracks which bits are unpredictable.
    When new_mask_flag is set, the mask is replaced with the build vector.

    Equation:
    - Mt = (It XOR It-1) OR Mt-1  (if new_mask_flag = 0)
    - Mt = (It XOR It-1) OR Bt-1  (if new_mask_flag = 1)

    Args:
        mask: Mask vector to update (modified in place)
        input_vec: Current input vector It
        prev_input: Previous input vector It-1
        build_prev: Previous build vector Bt-1
        new_mask_flag: Whether a new mask is being established
    """
    # Calculate changes: It XOR It-1
    changes = input_vec.copy()
    changes.xor(changes, prev_input)

    if new_mask_flag:
        # Mt = changes OR Bt-1
        mask.or_(changes, build_prev)
    else:
        # Mt = changes OR Mt-1
        mask.or_(mask, changes)


def compute_change(
    change: "BitVector",
    mask: "BitVector",
    prev_mask: "BitVector",
    t: int,
) -> None:
    """
    Compute the change vector (CCSDS Equation 8).

    The change vector tracks which mask bits changed between time steps.
    This is used in encoding to communicate mask updates.

    Equation:
    - Dt = Mt XOR Mt-1  (if t > 0)
    - Dt = Mt           (if t = 0, assuming M-1 = 0)

    Args:
        change: Change vector to compute (modified in place)
        mask: Current mask vector Mt
        prev_mask: Previous mask vector Mt-1
        t: Current time step
    """
    if t == 0:
        # At t=0, D0 = M0 (assuming M-1 = 0)
        change.copy_from(mask)
    else:
        # Dt = Mt XOR Mt-1
        change.xor(mask, prev_mask)

File: implementations/go/pocketplus/mask.go
package pocketplus

// UpdateBuild updates the build vector (CCSDS Equation 6).
//
// Build vector accumulates bits that have changed over time.
// When newMaskFlag is set, the build is used to replace the mask
// and is then reset to zero.
//
// Equation:
//   - Bt = (It XOR It-1) OR Bt-1  (if t > 0 and newMaskFlag = 0)
//   - Bt = 0                      (if t = 0 or newMaskFlag = 1)
func UpdateBuild(build, inputVec, prevInput *BitVector, newMaskFlag bool, t int) {
	if t == 0 || newMaskFlag {
		// Reset build to zero
		build.Zero()
	} else {
		// Bt = (It XOR It-1) OR Bt-1
		// Calculate changes
		changes := inputVec.XOR(prevInput)

		// Update build: build = changes OR build
		build.ORInto(changes, build)
	}
}

// UpdateMask updates the mask vector (CCSDS Equation 7).
//
// The mask tracks which bits are unpredictable.
// When newMaskFlag is set, the mask is replaced with the build vector.
//
// Equation:
//   - Mt = (It XOR It-1) OR Mt-1  (if newMaskFlag = 0)
//   - Mt = (It XOR It-1) OR Bt-1  (if newMaskFlag = 1)
func UpdateMask(mask, inputVec, prevInput, buildPrev *BitVector, newMaskFlag bool) {
	// Calculate changes: It XOR It-1
	changes := inputVec.XOR(prevInput)

	if newMaskFlag {
		// Mt = changes OR Bt-1
		mask.ORInto(changes, buildPrev)
	} else {
		// Mt = changes OR Mt-1
		result := changes.OR(mask)
		mask.CopyFrom(result)
	}
}

// ComputeChange computes the change vector (CCSDS Equation 8).
//
// The change vector tracks which mask bits changed between time steps.
// This is used in encoding to communicate mask updates.
//
// Equation:
//   - Dt = Mt XOR Mt-1  (if t > 0)
//   - Dt = Mt           (if t = 0, assuming M-1 = 0)
func ComputeChange(change, mask, prevMask *BitVector, t int) {
	if t == 0 {
		// At t=0, D0 = M0 (assuming M-1 = 0)
		change.CopyFrom(mask)
	} else {
		// Dt = Mt XOR Mt-1
		change.XORInto(mask, prevMask)
	}
}

// ComputeKt computes the kt component for encoding.
//
// kt = BE(Mt, Dt) - bits of mask at positions where change occurred
// This tells the decompressor which mask values to expect.
func ComputeKt(bb *BitBuffer, mask, change *BitVector) error {
	return BitExtractForward(bb, mask, change)
}

// ApplyPrediction applies prediction to get predicted value.
//
// Equation:
//   - P(It) = It-1 AND Mt (prediction based on mask)
//   - If mask bit is 1, predict same as previous
//   - If mask bit is 0, predict 0
func ApplyPrediction(prevInput, mask *BitVector) *BitVector {
	return prevInput.AND(mask)
}

// ComputeResidual computes residual (difference from prediction).
//
// Equation:
//   - Rt = It XOR P(It)
//   - Rt = It XOR (It-1 AND Mt)
func ComputeResidual(input, prevInput, mask *BitVector) *BitVector {
	prediction := ApplyPrediction(prevInput, mask)
	return input.XOR(prediction)
}

// ApplyReconstruction reconstructs input from residual and prediction.
//
// Equation:
//   - It = Rt XOR P(It)
//   - It = Rt XOR (It-1 AND Mt)
func ApplyReconstruction(residual, prevInput, mask *BitVector) *BitVector {
	prediction := ApplyPrediction(prevInput, mask)
	return residual.XOR(prediction)
}
